### PR TITLE
Story/x1683

### PR DESCRIPTION
**Pull request by kdwyer**
See also  #660, the UDES 11 version of this PR.  Ideally comments there will be addressed in both PRs.  Please comment here only if the issue is specific to the 14.0 branch.

### CODE REVIEW

File: addons/udes_stock/models/stock_picking_type.py [hunk 1]
```diff
@@ -70,6 +70,12 @@ class StockPickingType(models.Model):
         help="When True, operations are automatically validated in real time.",
     )
 
+    u_handle_partials = fields.Boolean(
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_picking_type.py:73]: Can we add this to the picking type view?

File: addons/udes_stock/models/stock_picking_type.py [hunk 2]
```diff
@@ -70,6 +70,12 @@ class StockPickingType(models.Model):
...
+        string="Process Partial Transfers",
+        default=True,
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_picking_type.py:75]: We'll need to review what picking types we have setup should have this set to False, not for this story though.

File: addons/udes_stock_cron/models/stock_picking.py [hunk 3]
```diff
@@ -21,6 +21,74 @@
...
+            #    prefetching, which leads to overcaching, described above
+            #  * cache misses for stock move fields result in hundreds of small
+            #    loads per batch, which is inefficient due to overheads
+            # For non-relational fields, read() is used instead of mapped()
+            # because it allows for loading of specific fields, whereas
+            # mapped() will load in as many fields as it can due to prefetching.
+            # with_context(prefetch_fields=False) could be used with mapped()
+            # but is limited to a single column at a time.
+            batch.with_context(prefetch_fields=False).mapped("move_lines")
+            batch.mapped("move_lines").read(
```
Comment (jrbirch99) [inline addons/udes_stock_cron/models/stock_picking.py:64]: Do we need the `mapped` here in udes 14?

File: addons/udes_stock_cron/models/stock_picking.py [hunk 4]
```diff
@@ -21,6 +21,74 @@
...
+            #  * cache misses cause unused fields to be loaded in due to
+            #    prefetching, which leads to overcaching, described above
+            #  * cache misses for stock move fields result in hundreds of small
+            #    loads per batch, which is inefficient due to overheads
+            # For non-relational fields, read() is used instead of mapped()
+            # because it allows for loading of specific fields, whereas
+            # mapped() will load in as many fields as it can due to prefetching.
+            # with_context(prefetch_fields=False) could be used with mapped()
+            # but is limited to a single column at a time.
+            batch.with_context(prefetch_fields=False).mapped("move_lines")
```
Comment (jrbirch99) [inline addons/udes_stock_cron/models/stock_picking.py:63]: Q: Why do you call `batch.with_context(prefetch_fields=False).mapped("move_lines")` first and not store it to a local variable, and then call `batch.mapped("move_lines").read(...)`?

File: addons/udes_stock/README.md [hunk 5]
```diff
@@ -187,6 +187,10 @@ The type of stock.picking can is defined by this type. It can represent a goods
 | u_scan_parent_package_end | Boolean | If the system should ask the user to scan a parent package on drop off                                                                   |
 | u_multi_users_enabled     | Boolean | Flag to enable multi users processing same picking simultaneously.                                                                       |
 | u_enable_unpickable_items | Boolean | Flag to indicate if the current picking type should support handling of unpickable items                                                 |
+| u_handle_partials | Boolean | If the picking type is allowed to handle
+partially available pickings. If True, then pickings of this type will report
+their u_pending value. If False, prevents the state of a picking to be ready
```
Comment (ws-mps) [inline addons/udes_stock/README.md:192]: Description needs updating.

File: addons/udes_stock/models/stock_quant.py [hunk 6]
```diff
@@ -101,3 +101,15 @@ def create_picking(self, picking_type, only_available=False, **kwargs):
         return Picking.with_context(quant_ids=self.ids).create_picking(
             picking_type, products_info, **kwargs
         )
+
+    @api.model
+    def get_available_quantity(self, product, locations):
+        """Get available quantity of product_id within locations."""
+        product.ensure_one()
+        domain = [("product_id", "=", product.id), ("location_id", "child_of", locations.ids)]
+        quants = self.search(domain).with_context(prefetch_fields=False)
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_quant.py:110]: Nit: `Quant.search()` please

File: addons/udes_stock_cron/models/stock_picking.py (hunk 4 above)
Comment (ws-mps) [reply addons/udes_stock_cron/models/stock_picking.py:63]: Agree, makes sense to put it in a variable.

File: addons/udes_stock_cron/models/stock_picking.py (hunk 3 above)
Comment (ws-mps) [reply addons/udes_stock_cron/models/stock_picking.py:64]: Mapped not needed for relational fields any more. We could use the variable mentioned above.

File: addons/udes_stock_cron/models/stock_picking.py [hunk 7]
```diff
@@ -116,6 +192,29 @@ def reserve_stock(self):
                 if batch and picking_type.u_reserve_batches:
                     pickings = batch.picking_ids
 
+                # Check for available stock and add pickings which we cannot
+                # fulfill to the processed recordset. The logic here
+                # should mirror that within the savepoint below, expect here we
+                # do not have the move lines.
+                unreservable_moves = pickings._find_unreservable_moves(return_fast=bool(self))
+                if unreservable_moves:
+                    moves = pickings.mapped("move_lines")
```
Comment (jrbirch99) [inline addons/udes_stock_cron/models/stock_picking.py:201]: nit: mapped not needed

File: addons/udes_stock/models/stock_picking_type.py (hunk 2 above)
Comment (kdwyer) [reply addons/udes_stock/models/stock_picking_type.py:75]: Yep - agreed with @ws-mps that we would just default to true for ow.

File: addons/udes_stock/README.md (hunk 5 above)
Comment (kdwyer) [reply addons/udes_stock/README.md:192]: I can truncate from "If True..." - or do you want some additional text?

File: addons/udes_stock/models/stock_quant.py (hunk 6 above)
Comment (kdwyer) [reply addons/udes_stock/models/stock_quant.py:110]: How do we get `Quant` here?  Do we need to do `Quant = self.env[...]`?  Or use the class name (`StockQuant`)?

File: addons/udes_stock/models/stock_picking_type.py (hunk 1 above)
Comment (kdwyer) [reply addons/udes_stock/models/stock_picking_type.py:73]: Added, please review

File: addons/udes_stock_cron/models/stock_picking.py (hunk 4 above)
Comment (kdwyer) [reply addons/udes_stock_cron/models/stock_picking.py]: Changed, please review.

File: addons/udes_stock_cron/models/stock_picking.py (hunk 7 above)
Comment (kdwyer) [reply addons/udes_stock_cron/models/stock_picking.py]: Changed, please review.

File: addons/udes_stock/models/stock_quant.py (hunk 6 above)
Comment (kdwyer) [reply addons/udes_stock/models/stock_quant.py:110]: hmm, seems we have to go down the env route...

**Review by jrbirch99 — APPROVED** [review]

File: addons/udes_stock/models/stock_quant.py (hunk 6 above)
Comment (kdwyer) [reply addons/udes_stock/models/stock_quant.py]: Changed, please review.

**Review by ws-mps — APPROVED** [review]

**Review by caleb-shelton — APPROVED** [review]

**Review by PeterC10 — APPROVED** [review]

File: addons/udes_stock/wizard/reservation_views.xml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    </record>
+
+    <record id="stock_reserve_stock_action_stock_picking" model="ir.actions.act_window">
+        <field name="name">Reserve Stock</field>
+        <field name="res_model">stock.reserve.stock.wizard</field>
+        <field name="view_mode">form</field>
+        <field name="target">new</field>
+        <field name="binding_model_id" ref="model_stock_picking"/>
+        <field name="binding_view_types">form,list</field>
+      </record>
```
Comment (armandcela) [inline addons/udes_stock/wizard/reservation_views.xml:33]: Nit: Formatting

**Review by ws-mps — APPROVED** [review]

**Review by armandcela — APPROVED** [review]

**Pull request merged**